src/applications/gi/reducers/calculator.js

```diff
@@ -69,6 +76,91 @@ export default function (state = INITIAL_STATE, action) {
         ...newState
       };
     }
+    case FETCH_BAH_FAILED: {
+      const { beneficiaryZipFetched } = action;
+      const { error } = action.payload;
+
+      // response mismatch - do nothing
+      if (beneficiaryZipFetched !== state.beneficiaryZipFetched) {
+        return state;
+      }
+
+      const newState = {
+        beneficiaryZipError: error,
+        beneficiaryZipFetched: '',
+        beneficiaryLocationBah: null,
+        housingAllowanceCity: ''
+      };
+
+      return {
+        ...state,
+        ...newState
+      };
+    }
+
+    case FETCH_BAH_STARTED: {
+      const { beneficiaryZipFetched } = action;
+
+      const newState = {
+        beneficiaryZipError: '',
+        beneficiaryZip: beneficiaryZipFetched,
+        beneficiaryZipFetched,
+        housingAllowanceCity: 'Loading...'
+      };
+
+      return {
+        ...state,
+        ...newState
+      };
+    }
+
+    case FETCH_BAH_SUCCEEDED: {
+      const { beneficiaryZipFetched } = action;
+      const {
+        bah: beneficiaryLocationBah,
+        city: housingAllowanceCity } = action.payload;
+
+      // response mismatch - do nothing
+      if (beneficiaryZipFetched !== state.beneficiaryZipFetched) {
+        return state;
+      }
+
+      const newState = {
+        beneficiaryZipError: '',
+        beneficiaryZipFetched: '',
+        beneficiaryLocationBah,
+        housingAllowanceCity
+      };
+
+      return {
+        ...state,
+        ...newState
+      };
+    }
+    case BENEFICIARY_ZIP_CODE_CHANGED: {
+      const { beneficiaryZip } = action;
+
+      let beneficiaryZipError = state.beneficiaryZipError || '';
+
+      if (beneficiaryZip.length >= 5 && !beneficiaryZipRegExTester.exec(beneficiaryZip)) {
```

Comment:
Do you need both of these? Could you write the regex to require that it's 5 digits?